include/swift/SILOptimizer/Utils/Differentiation/Common.h

```diff
@@ -51,6 +58,70 @@ template <class Inst> Inst *peerThroughFunctionConversions(SILValue value) {
   return nullptr;
 }
 
+/// Given a range of elements, joins these into a single value. If there's
+/// exactly one element, returns that element. Otherwise, creates a tuple using
+/// a `tuple` instruction.
+SILValue joinElements(ArrayRef<SILValue> elements, SILBuilder &builder,
+                      SILLocation loc);
+
+/// Given a value, extracts all elements to `results` from this value if it has
+/// a tuple type. Otherwise, add this value directly to `results`.
+void extractAllElements(SILValue value, SILBuilder &builder,
+                        SmallVectorImpl<SILValue> &results);
+
+/// Emit a zero value into the given buffer access by calling
+/// `AdditiveArithmetic.zero`. The given type must conform to
+/// `AdditiveArithmetic`.
+void emitZeroIntoBuffer(SILBuilder &builder, CanType type,
```

Comment:
This change seems important for correctness. It can probably help clean up this code:
https://github.com/apple/swift/blob/13d5a8addbe3605984edc4ce7c6cbf1f0649a9e1/lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp#L486-L497

I started looking into this but ran into some difficulties. The user above wants to insert at the end of a basic block (`SILBasicBlock::iterator`), but no appropriate constructor exists. Only `SILBuilderWithScope(SILInstruction *I, SILBuilderContext &C)` is recommended.

`SILBuilderWithScope(SILBasicBlock::iterator I, SILBuilderContext &C)` would work but doesn't exist. Can we add it, or are there other considerations?